### PR TITLE
Refactor Game class to extract GameUI

### PR DESCRIPTION
**Pull request by nerzhul**
GameUI refactor (part 1/X): GameUI object creation + GameUIFlags move…

Game class is too huge and has too specialization on various subjects, like UI, formspecs, client, renderer. Start to move UI related things to GameUI object and cleanup them

Other improvements:
* updateChat: more performance on error messages by remove string copies
* Initialize all game class members in definition instead of constructor (with nullptr instead of NULL)
* Drop unused Client::show{GameChat,GameHud,Profiler,GameFog}

### CODE REVIEW

**Comment by rubenwardy** [comment]
I like any attempt to clean Game, but it must be done carefully and cleanly. I'd appreciate you giving any plans for further splitting up

**Comment by nerzhul** [comment]
This will be done step by step in a non squashed patchset (except intermediate fixes), and carefully, with new unittests additions as i want to have more MT coverage across this. Another goal i have in the long term is to split client & UI thread in 2 threads

**Comment by nerzhul** [comment]
@rubenwardy ready for a merge

**Comment by nerzhul** [comment]
This pr must not be squashed. Please also read each commit log to see changes separately

File: src/client/gameui.h [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+	gui::IGUIStaticText *m_guitext_info = nullptr; // At the middle of the screen
+	std::wstring m_infotext;
+
+	gui::IGUIStaticText *m_guitext_status = nullptr;
+	std::wstring m_statustext;
+	float m_statustext_time = 0.0f;
+
+	gui::IGUIStaticText *m_guitext_chat; // Chat text
+
+	gui::IGUIStaticText *m_guitext_profiler; // Profiler text
```
Comment (SmallJoker) [inline src/client/gameui.h]: Please initialize these two pointers to `nullptr` too.

File: src/game.cpp [hunk 2]
```diff
@@ -1388,45 +1274,47 @@ class Game {
 		CameraOrientation *cam);
 	void handleClientEvent_CloudParams(ClientEvent *event, CameraOrientation *cam);
 
+	void updateChat(f32 dtime, const v2u32 &screensize);
 	static const ClientEventHandler clientEventHandler[CLIENTEVENT_MAX];
 
 	InputHandler *input;
```
Comment (SmallJoker) [inline src/game.cpp]: `= nullptr`

File: src/game.cpp [hunk 3]
```diff
@@ -2726,20 +2532,17 @@ void Game::toggleFreeMove()
 	bool free_move = !g_settings->getBool("free_move");
 	g_settings->set("free_move", bool_to_cstr(free_move));
 
-	runData.statustext_time = 0;
-
 	if (free_move) {
 		if (client->checkPrivilege("fly")) {
-			showStatusTextSimple("Fly mode enabled");
+			m_game_ui->showTranslatedStatusText("Fly mode enabled");
```
Comment (SmallJoker) [inline src/game.cpp]: Needs to be added to our translation updating script, so these texts are detected.

File: src/game.cpp [hunk 4]
```diff
@@ -2856,47 +2633,45 @@ void Game::toggleMinimap(bool shift_pressed)
...
-	runData.statustext_time = 0;
-        if (flags.force_fog_off)
-                showStatusTextSimple("Fog disabled");
-        else
-                showStatusTextSimple("Fog enabled");
+	m_flags.force_fog_off = !m_flags.force_fog_off;
+	if (m_flags.force_fog_off)
+			m_game_ui->showTranslatedStatusText("Fog disabled");
+	else
+			m_game_ui->showTranslatedStatusText("Fog enabled");
```
Comment (SmallJoker) [inline src/game.cpp]: Too many indents.

File: src/game.cpp [hunk 5]
```diff
@@ -2906,64 +2681,38 @@ void Game::toggleDebug()
...
-		swprintf(buf, sizeof(buf) / sizeof(wchar_t), str,
-				runData.profiler_current_page,
-				runData.profiler_max_page);
-		delete[] str;
-		m_statustext = buf;
-	} else {
-		showStatusTextSimple("Profiler hidden");
-	}
-	runData.statustext_time = 0;
+		m_game_ui->showTranslatedStatusText("Camera update enabled");
```
Comment (SmallJoker) [inline src/game.cpp]: Too many indents.

File: src/game.cpp [hunk 6]
```diff
@@ -3166,12 +2912,10 @@ inline void Game::step(f32 *dtime)
 	if (can_be_and_is_paused) {	// This is for a singleplayer server
 		*dtime = 0;             // No time passes
 	} else {
-		if (server != NULL) {
-			//TimeTaker timer("server->step(dtime)");
+		if (server) {
```
Comment (SmallJoker) [inline src/game.cpp]: Brace removal

File: src/gui/guiFormSpecMenu.cpp [hunk 7]
```diff
@@ -2024,7 +2025,8 @@ void GUIFormSpecMenu::regenerateGui(v2u32 screensize)
 	{
 		assert(!m_tooltip_element);
 		// Note: parent != this so that the tooltip isn't clipped by the menu rectangle
-		m_tooltip_element = addStaticText(Environment, L"",core::rect<s32>(0,0,110,18));
+		m_tooltip_element = gui::StaticText::add(Environment, L"",
+			core::rect<s32>(0,0,110,18));
```
Comment (SmallJoker) [inline src/gui/guiFormSpecMenu.cpp]: Spaces after each comma

File: src/irrlicht_changes/static_text.h [hunk 8]
```diff
@@ -42,6 +42,59 @@ namespace gui
...
+			irr::gui::IGUIEnvironment *guienv,
+			const wchar_t *text,
+			const core::rect< s32 > &rectangle,
+			bool border = false,
+			bool wordWrap = true,
+			irr::gui::IGUIElement *parent = NULL,
+			s32 id = -1,
+			bool fillBackground = false)
+		{
+			return add(guienv, EnrichedString(text), rectangle, border, wordWrap, parent, id, fillBackground);
```
Comment (SmallJoker) [inline src/irrlicht_changes/static_text.h]: Line too long. Function parameter `parent` should default to `nullptr`

File: src/game.cpp (hunk 3 above)
Comment (nerzhul) [reply src/game.cpp]: done

File: src/game.cpp (hunk 5 above)
Comment (nerzhul) [reply src/game.cpp]: Fixed

**Comment by nerzhul** [comment]
approved by @SmallJoker http://irc.minetest.net/minetest-dev/2018-01-05#i_5190878

**Pull request merged**